File: ios-app/CYDDash/Services/TollEngine.swift
// Time-varying toll engine — mirror of firmware/src/model/toll_engine.cpp.
// Schedule source of truth: docs/toll-schedule.md. Parity is enforced by
// CYDDashTests.TollEngineTests against the shared test vectors.
import Foundation

enum TollEngine {
    enum Crossing: CaseIterable {
        case whc, cht, ehc
        var name: String {
            switch self {
            case .whc: return "西隧"
            case .cht: return "紅隧"
            case .ehc: return "東隧"
            }
        }
    }

    struct Result: Equatable {
        let dollars: Int
        let nextDollars: Int
        let nextChangeSec: Int  // 86400 = no further change today
    }

    // toll(t) = v0 + dir * 2 * floor((t - from) / 120); dir 0 = plateau
    private struct Seg {
        let from: Int
        let to: Int
        let v0: Int
        let dir: Int
    }

    private static func s(_ h: Int, _ m: Int) -> Int { h * 3600 + m * 60 }

    private static let profileW: [Seg] = [
        .init(from: 0, to: s(7, 30), v0: 20, dir: 0),
        .init(from: s(7, 30), to: s(8, 8), v0: 22, dir: 1),
        .init(from: s(8, 8), to: s(10, 15), v0: 60, dir: 0),
        .init(from: s(10, 15), to: s(10, 43), v0: 58, dir: -1),
        .init(from: s(10, 43), to: s(16, 30), v0: 30, dir: 0),
        .init(from: s(16, 30), to: s(16, 58), v0: 32, dir: 1),
        .init(from: s(16, 58), to: s(19, 0), v0: 60, dir: 0),
        .init(from: s(19, 0), to: s(19, 38), v0: 58, dir: -1),
        .init(from: s(19, 38), to: 86400, v0: 20, dir: 0),
    ]

    private static let profileC: [Seg] = [
        .init(from: 0, to: s(7, 30), v0: 20, dir: 0),
        .init(from: s(7, 30), to: s(7, 48), v0: 22, dir: 1),
        .init(from: s(7, 48), to: s(10, 15), v0: 40, dir: 0),
        .init(from: s(10, 15), to: s(10, 23), v0: 38, dir: -1),
        .init(from: s(10, 23), to: s(16, 30), v0: 30, dir: 0),
        .init(from: s(16, 30), to: s(16, 38), v0: 32, dir: 1),
        .init(from: s(16, 38), to: s(19, 0), v0: 40, dir: 0),
        .init(from: s(19, 0), to: s(19, 18), v0: 38, dir: -1),
        .init(from: s(19, 18), to: 86400, v0: 20, dir: 0),
    ]

    private static let profileS: [Seg] = [
        .init(from: 0, to: s(10, 11), v0: 20, dir: 0),
        .init(from: s(10, 11), to: s(10, 13), v0: 21, dir: 0),
        .init(from: s(10, 13), to: s(10, 15), v0: 23, dir: 0),
        .init(from: s(10, 15), to: s(19, 15), v0: 25, dir: 0),
        .init(from: s(19, 15), to: s(19, 17), v0: 23, dir: 0),
        .init(from: s(19, 17), to: s(19, 19), v0: 21, dir: 0),
        .init(from: s(19, 19), to: 86400, v0: 20, dir: 0),
    ]

    private static func profile(_ c: Crossing, sundayOrPH: Bool) -> [Seg] {
        if sundayOrPH { return profileS }
        return c == .whc ? profileW : profileC
    }

    private static func eval(_ seg: Seg, _ t: Int) -> Int {
        seg.dir == 0 ? seg.v0 : seg.v0 + seg.dir * 2 * ((t - seg.from) / 120)
    }

    private static func evalAt(_ segs: [Seg], _ t: Int) -> Int {
        segs.first { t >= $0.from && t < $0.to }.map { eval($0, t) } ?? 20
    }

    static func query(_ c: Crossing, secOfDay: Int, sundayOrPH: Bool) -> Result {
        let t = secOfDay % 86400
        let segs = profile(c, sundayOrPH: sundayOrPH)
        guard let seg = segs.first(where: { t >= $0.from && t < $0.to }) else {
            return Result(dollars: 20, nextDollars: 20, nextChangeSec: 86400)
        }
        let dollars = eval(seg, t)

        var next = seg.to
        if seg.dir != 0 {
            let step = seg.from + ((t - seg.from) / 120 + 1) * 120
            if step < seg.to { next = step }
        }
        if next >= 86400 {
            return Result(dollars: dollars, nextDollars: dollars, nextChangeSec: 86400)
        }
        var nextChange = next
        var nextDollars = evalAt(segs, next)
        // Skip value-neutral boundaries (plateau -> equal ramp start)
        while nextChange < 86400, nextDollars == dollars {
            let deeper = query(c, secOfDay: nextChange, sundayOrPH: sundayOrPH)
            nextChange = deeper.nextChangeSec
            nextDollars = deeper.nextDollars
        }
        return Result(dollars: dollars, nextDollars: nextDollars, nextChangeSec: nextChange)
    }
}

File: firmware/src/model/toll_engine.cpp
#include "toll_engine.h"

#include <cstddef>

namespace toll {
namespace {

// toll(t) = v0 + dir * 2 * floor((t - from) / 120), dir=0 => plateau
struct Seg {
  uint32_t from;  // inclusive, sec of day
  uint32_t to;    // exclusive
  uint8_t v0;
  int8_t dir;
};

constexpr uint32_t S(uint32_t h, uint32_t m, uint32_t s = 0) { return h * 3600 + m * 60 + s; }

// WHC 西隧, Mon-Sat excl. PH
constexpr Seg kProfileW[] = {
    {S(0, 0),   S(7, 30),  20, 0},
    {S(7, 30),  S(8, 8),   22, +1},  // -> $58
    {S(8, 8),   S(10, 15), 60, 0},
    {S(10, 15), S(10, 43), 58, -1},  // -> $32
    {S(10, 43), S(16, 30), 30, 0},
    {S(16, 30), S(16, 58), 32, +1},  // -> $58
    {S(16, 58), S(19, 0),  60, 0},
    {S(19, 0),  S(19, 38), 58, -1},  // -> $22
    {S(19, 38), 86400,     20, 0},
};

// CHT 紅隧 & EHC 東隧, Mon-Sat excl. PH
constexpr Seg kProfileC[] = {
    {S(0, 0),   S(7, 30),  20, 0},
    {S(7, 30),  S(7, 48),  22, +1},  // -> $38
    {S(7, 48),  S(10, 15), 40, 0},
    {S(10, 15), S(10, 23), 38, -1},  // -> $32
    {S(10, 23), S(16, 30), 30, 0},
    {S(16, 30), S(16, 38), 32, +1},  // -> $38
    {S(16, 38), S(19, 0),  40, 0},
    {S(19, 0),  S(19, 18), 38, -1},  // -> $22
    {S(19, 18), 86400,     20, 0},
};

// Sundays & public holidays, all three crossings
constexpr Seg kProfileS[] = {
    {S(0, 0),   S(10, 11), 20, 0},
    {S(10, 11), S(10, 13), 21, 0},
    {S(10, 13), S(10, 15), 23, 0},
    {S(10, 15), S(19, 15), 25, 0},
    {S(19, 15), S(19, 17), 23, 0},
    {S(19, 17), S(19, 19), 21, 0},
    {S(19, 19), 86400,     20, 0},
};

struct Profile {
  const Seg* segs;
  size_t n;
};

Profile profileFor(Crossing c, bool sundayOrPH) {
  if (sundayOrPH) return {kProfileS, sizeof(kProfileS) / sizeof(Seg)};
  if (c == Crossing::WHC) return {kProfileW, sizeof(kProfileW) / sizeof(Seg)};
  return {kProfileC, sizeof(kProfileC) / sizeof(Seg)};
}

uint8_t evalSeg(const Seg& s, uint32_t t) {
  if (s.dir == 0) return s.v0;
  return static_cast<uint8_t>(s.v0 + s.dir * 2 * static_cast<int>((t - s.from) / 120));
}

uint8_t evalAt(const Profile& p, uint32_t t) {
  for (size_t i = 0; i < p.n; i++)
    if (t >= p.segs[i].from && t < p.segs[i].to) return evalSeg(p.segs[i], t);
  return p.segs[p.n - 1].v0;  // unreachable for t < 86400
}

}  // namespace

Result query(Crossing c, uint32_t secOfDay, bool sundayOrPH) {
  if (secOfDay >= 86400) secOfDay %= 86400;
  const Profile p = profileFor(c, sundayOrPH);

  for (size_t i = 0; i < p.n; i++) {
    const Seg& s = p.segs[i];
    if (secOfDay < s.from || secOfDay >= s.to) continue;

    Result r;
    r.dollars = evalSeg(s, secOfDay);

    // Next change: next ramp step within this segment, else the segment boundary.
    uint32_t next = s.to;
    if (s.dir != 0) {
      uint32_t step = s.from + ((secOfDay - s.from) / 120 + 1) * 120;
      if (step < s.to) next = step;
    }
    if (next >= 86400) {
      r.next_change_sec = 86400;
      r.next_dollars = r.dollars;  // both schedules end and start the day at $20
    } else {
      r.next_change_sec = next;
      r.next_dollars = evalAt(p, next);
    }
    // Boundary may be value-neutral (e.g. plateau -> equal ramp start); skip ahead.
    while (r.next_change_sec < 86400 && r.next_dollars == r.dollars) {
      Result deeper = query(c, r.next_change_sec, sundayOrPH);
      r.next_change_sec = deeper.next_change_sec;
      r.next_dollars = deeper.next_dollars;
    }
    return r;
  }
  return {20, 20, 86400};
}

}  // namespace toll
